tests/integration/pilot/upgrade/multi_version_revision_test.go

```diff
@@ -0,0 +1,167 @@
+// should test all possible revisioned namespace pairings to test traffic between all versions
+func TestMultiVersionRevision(t *testing.T) {
+	framework.NewTest(t).
+		RequiresSingleCluster().
+		Features("installation.upgrade").
+		Run(func(ctx framework.TestContext) {
+			goldenDir := filepath.Join(env.IstioSrc, "tests/integration/pilot/upgrade/goldens")
+
+			// keep these at the latest patch version of each minor version
+			// TODO(samnaser) script the creation of these golden files
+			installGoldens := []installGolden{
+				{
+					revision: "1-7-4",
+					path:     filepath.Join(goldenDir, "golden-minimal-1.7.4.yaml"),
+				},
+				{
+					revision: "1-6-11",
+					path:     filepath.Join(goldenDir, "golden-minimal-1.6.11.yaml"),
+				},
+			}
+
+			// keep track of applied configurations and clean up after the test
+			configs := make(map[string]string)
+			defer func() {
+				for _, config := range configs {
+					ctx.Config().DeleteYAML("istio-system", config)
+				}
+			}()
+
+			nsNoRevision := namespace.NewOrFail(t, ctx, namespace.Config{
+				Prefix: "no-revision",
+				Inject: true,
+			})
+			echoNamespaces := []revisionedNamespace{
+				{
+					revision:  "none",
+					namespace: nsNoRevision,
+				},
+			}
+
+			for _, g := range installGoldens {
+				configBytes, err := ioutil.ReadFile(g.path)
+				if err != nil {
+					t.Errorf("failed to read config golden from path %s: %v", g.path, err)
+				}
+				configs[g.revision] = string(configBytes)
+				ctx.Config().ApplyYAMLOrFail(t, "istio-system", string(configBytes))
+
+				// create a namespace pointing to each revisioned install
+				ns := namespace.NewOrFail(t, ctx, namespace.Config{
+					Prefix:   fmt.Sprintf("revision-%s", g.revision),
+					Inject:   true,
+					Revision: g.revision,
+				})
+				echoNamespaces = append(echoNamespaces, revisionedNamespace{
+					revision:  g.revision,
+					namespace: ns,
+				})
+
+				// need to sleep a few seconds or else both installs won't succeed?
+				time.Sleep(time.Second * 2)
+			}
+
+			// create an echo instance in each revisioned namespace, all these echo
+			// instances will be injected with proxies from their respective versions
+			builder := echoboot.NewBuilder(ctx)
+			instances := make([]echo.Instance, len(echoNamespaces))
+			for i, ns := range echoNamespaces {
+				builder = builder.With(&instances[i], echo.Config{
+					Service:   fmt.Sprintf("revision-%s", ns.revision),
+					Namespace: ns.namespace,
+					Ports: []echo.Port{
+						{
+							Name:         "http",
+							Protocol:     protocol.HTTP,
+							InstancePort: 8080,
+						}},
+				})
+			}
+			builder.BuildOrFail(t)
+			testAllEchoCalls(t, instances)
+		})
+}
+
+// generateEchoCalls takes list of revisioned namespaces and generates list of echo calls covering
+// communication between every pair of namespaces
+func testAllEchoCalls(t *testing.T, echoInstances []echo.Instance) {
+	for i := 0; i < len(echoInstances); i++ {
```

Comment:
I think you can just make this `for_, source := range echoInstances`. Or `for i := range echoInstances` if you need the index, but I don't think you do

Reply:
Ah good point, changed